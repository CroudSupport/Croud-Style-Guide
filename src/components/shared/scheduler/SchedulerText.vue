<template>
    <div>
        <span v-if="ended">Ran at</span>
        <span v-else>Runs at</span>

        <strong>{{ schedule.frequency.at }}</strong>

        <span v-html="daysText"></span>

        <span v-if="periodText">
            on <strong>{{ periodText }}</strong>
        </span>

        <span v-html="monthsText"></span>

        <span v-html="datesText"></span>
    </div>
</template>

<script>
    import moment from 'moment'
    import { difference } from 'lodash'
    import schedule from '../mixins/schedule'

/**
* Text representation of a schedule
*
* @example ./croud-scheduler-text.md
*/
export default {
        name: 'croud-scheduler-text',

        mixins: [schedule],

        props: {
            /**
             * Root object to contain scheulder schema
             *
            */
            rootObject: {
                type: Object,
                required: true,
            },
        },

        data() {
            return {
                weekdays: moment.weekdays().slice(1, -1),
                dateFormat: 'll',
                ended: false,
                periods: {
                    daily: false,
                },
            }
        },

        computed: {
            days() {
                return this.schedule.frequency.days
            },

            months() {
                return this.schedule.frequency.months
            },

            daysText() {
                const days = Object.keys(this.days).filter(key => this.days[key]).map(key => `${key[0].toUpperCase()}${key.substr(1)}`)

                if (!days.length || days.length === 7) {
                    return '<strong>every day</strong>'
                }

                if (difference(this.weekdays, days).length === 0) {
                    return 'on <strong>weekdays</strong>'
                }

                return `on <strong>${this.joinText(days)}</strong>`
            },

            monthsText() {
                const months = Object.keys(this.months).filter(key => this.months[key]).map(key => `${key[0].toUpperCase()}${key.substr(1)}`)

                if (!months.length || months.length === 12) {
                    return ''
                }

                return `during <strong>${this.joinText(months)}</strong>`
            },

            periodText() {
                return this.periods[this.schedule.frequency.recur]
            },

            datesText() {
                this.ended = false
                const today = moment()

                const startsAt = this.schedule.limit.startsAt ? moment(this.schedule.limit.startsAt) : null
                const endsAt = this.schedule.limit.endsAt ? moment(this.schedule.limit.endsAt) : null

                if (startsAt && startsAt.isAfter(today) && endsAt && endsAt.isAfter(today)) {
                    return `between <strong>${startsAt.format(this.dateFormat)}</strong> & <strong>${endsAt.format(this.dateFormat)}</strong>`
                }

                if (startsAt && startsAt.isAfter(today)) {
                    return `starting from <strong>${startsAt.format(this.dateFormat)}</strong>`
                }

                if (endsAt && endsAt.isAfter(today)) {
                    return `ending on <strong>${endsAt.format(this.dateFormat)}</strong>`
                }

                if (endsAt && endsAt.isBefore(today)) {
                    this.ended = true
                    return `but ended on <strong>${endsAt.format(this.dateFormat)}</strong>`
                }

                return ''
            },
        },

        methods: {
            joinText(arr, joiner = ', ', finalJoiner = ' & ') {
                return [
                    arr.slice(0, -1).join(joiner),
                    arr.slice(-1)[0],
                ].join(arr.length < 2 ? '' : finalJoiner)
            },
        },

        created() {
            this.getSchedule()
        },

        watch: {
            rootObject: 'getSchedule',
        },
    }
</script>
