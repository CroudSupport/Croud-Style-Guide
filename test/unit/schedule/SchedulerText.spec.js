import Vue from 'vue'
import moment from 'moment'
import VueSemantic from 'croud-vue-semantic'
import SchedulerText from '../../../src/components/shared/scheduler/SchedulerText'

Vue.use(VueSemantic)

const Constructor = Vue.extend(SchedulerText)

const keys = {
    frequency: 'service=scheduler;table=timetables;field=frequency;',
    startsAt: 'service=scheduler;table=timetables;field=starts_at;',
    endsAt: 'service=scheduler;table=timetables;field=ends_at;',
    maxExecutions: 'service=scheduler;table=timetables;field=max_executions;',
}

const noScheduleVm = new Constructor({
    propsData: {
        rootObject: { id: 1, name: 'rootObject' },
    },
}).$mount()

const vm = new Constructor({
    propsData: {
        rootObject: {
            id: 1,
            name: 'rootObject',
            [keys.frequency]: {
                recur: 'everyFortnight',
                at: ['00:00'],
                timezone: 'Europe/London',
                months: ['march', 'april', 'may'],
                days: ['monday'],
            },
            [keys.maxExecutions]: 10,
            [keys.startsAt]: '2018-03-07 12:00:00',
            [keys.endsAt]: '2018-03-08 12:00:00',
        },
    },
}).$mount()

describe('Scheduler Text', () => {
    describe('build and utils', () => {
        it('set up with mixin', () => {
            expect(typeof vm.schedule).toBe('object')
            expect(typeof vm.getSchedule).toBe('function')
        })
        it('should join text', () => {
            expect(vm.joinText(['foo', 'bar', 'baz'])).toBe('foo, bar & baz')
        })
    })

    describe('no schedule data', () => {
        let text = ''

        it('should match the snapshot', () => {
            text = noScheduleVm.$el.textContent
            expect(text).toMatchSnapshot()
        })

        it('should react to the prop change', (done) => {
            noScheduleVm.rootObject = {
                [keys.frequency]: {
                    days: ['friday'],
                },
            }
            noScheduleVm.$nextTick(() => {
                expect(noScheduleVm.$el.textContent).not.toBe(text)
                done()
            })
        })
    })

    describe('existing schedule', () => {
        it('should match the snapshot', () => {
            expect(vm.$el.textContent).toMatchSnapshot()
        })

        describe('days', () => {
            it('should handle a list', (done) => {
                vm.rootObject = {
                    [keys.frequency]: {
                        days: ['monday', 'tuesday', 'wednesday'],
                    },
                }

                vm.$nextTick(() => {
                    expect(vm.daysText).toBe('on <strong>Monday, Tuesday & Wednesday</strong>')
                    done()
                })
            })

            it('should notice all weekdays', (done) => {
                vm.rootObject = {
                    [keys.frequency]: {
                        days: ['monday', 'tuesday', 'wednesday', 'thursday', 'friday'],
                    },
                }

                vm.$nextTick(() => {
                    expect(vm.daysText).toBe('on <strong>weekdays</strong>')
                    done()
                })
            })
        })

        describe('limits', () => {
            it('should handle an upcoming date range', (done) => {
                vm.rootObject = {
                    [keys.startsAt]: moment().add(1, 'day').format('YYYY-MM-DD HH:mm:ss'),
                    [keys.endsAt]: moment().add(2, 'day').format('YYYY-MM-DD HH:mm:ss'),
                }
                vm.$nextTick(() => {
                    expect(vm.datesText).toContain('between')
                    done()
                })
            })

            describe('end date', () => {
                it('should handle end date', (done) => {
                    vm.rootObject = {
                        [keys.startsAt]: moment().subtract(1, 'day').format('YYYY-MM-DD HH:mm:ss'),
                        [keys.endsAt]: moment().add(1, 'day').format('YYYY-MM-DD HH:mm:ss'),
                    }
                    vm.$nextTick(() => {
                        expect(vm.datesText).toContain('ending')
                        expect(vm.ended).toBe(false)
                        done()
                    })
                })

                it('should handle passed end date', (done) => {
                    vm.rootObject = {
                        [keys.startsAt]: moment().subtract(2, 'day').format('YYYY-MM-DD HH:mm:ss'),
                        [keys.endsAt]: moment().subtract(1, 'day').format('YYYY-MM-DD HH:mm:ss'),
                    }
                    vm.$nextTick(() => {
                        expect(vm.datesText).toContain('ended')
                        expect(vm.ended).toBe(true)
                        done()
                    })
                })
            })
        })
    })
})
