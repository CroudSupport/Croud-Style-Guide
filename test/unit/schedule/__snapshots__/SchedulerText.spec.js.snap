// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scheduler Text existing schedule should match the snapshot 1`] = `
"Ran at 00:00 on Monday 
        on Every alternating week during March, April & May but ended on Mar 8, 2018"
`;

exports[`Scheduler Text no schedule data should match the snapshot 1`] = `"Runs at 00:00 every day   "`;
